letsencrypt-nginx/letsencrypt_nginx/nginxparser.py

```diff
@@ -1,56 +1,241 @@
-"""Very low-level nginx config parser based on pyparsing."""
-import string
+"""Very low-level hand rolled nginx config parser, using recursive descent to
+retain all parsed information, so that the file can be reconstituted."""
 
-from pyparsing import (
-    Literal, White, Word, alphanums, CharsNotIn, Forward, Group,
-    Optional, OneOrMore, Regex, ZeroOrMore)
-from pyparsing import stringEnd
-from pyparsing import restOfLine
+import StringIO
 
+class ParseException(Exception):
+    """A parsing exception happened."""
+    def __init__(self, value):
+        self.value = value
+        super(ParseException, self).__init__()
+
+    def __str__(self):
+        return repr(self.value)
 
 class RawNginxParser(object):
     # pylint: disable=expression-not-assigned
-    """A class that parses nginx configuration with pyparsing."""
-
-    # constants
-    left_bracket = Literal("{").suppress()
-    right_bracket = Literal("}").suppress()
-    semicolon = Literal(";").suppress()
-    space = White().suppress()
-    key = Word(alphanums + "_/")
-    # Matches anything that is not a special character AND any chars in single
-    # or double quotes
-    value = Regex(r"((\".*\")?(\'.*\')?[^\{\};,]?)+")
-    location = CharsNotIn("{};," + string.whitespace)
-    # modifier for location uri [ = | ~ | ~* | ^~ ]
-    modifier = Literal("=") | Literal("~*") | Literal("~") | Literal("^~")
-
-    # rules
-    comment = Literal('#') + restOfLine()
-    assignment = (key + Optional(space + value, default=None) + semicolon)
-    location_statement = Optional(space + modifier) + Optional(space + location)
-    if_statement = Literal("if") + space + Regex(r"\(.+\)") + space
-    block = Forward()
-
-    block << Group(
-        (Group(key + location_statement) ^ Group(if_statement)) +
-        left_bracket +
-        Group(ZeroOrMore(Group(comment | assignment) | block)) +
-        right_bracket)
-
-    script = OneOrMore(Group(comment | assignment) ^ block) + stringEnd
+    """A class that parses nginx configuration."""
 
     def __init__(self, source):
         self.source = source
 
-    def parse(self):
-        """Returns the parsed tree."""
-        return self.script.parseString(self.source)
-
     def as_list(self):
         """Returns the parsed tree as a list."""
-        return self.parse().asList()
+        return self.parse()
 
+    def parse(self):
+        """Returns the parsed tree."""
+        return RawNginxParser.Internal(self.source).parse_file()
+
+    class Internal(object):
+        """An internal class to keep track of parsing state."""
+        def __init__(self, source):
+            self.src = StringIO.StringIO(source)
```

Comment:
Yeah, this is probably a good idea. I did it initially because I assumed the parser would want to work on file objects - what happens if we encounter a larger configuration file? But maybe just using the string directly is fine.